File: app/src/main/java/com/lotrapp/sander/lotrscenariogenerator/MainActivity.java
package com.lotrapp.sander.lotrscenariogenerator;

import android.content.Intent;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Random;

public class MainActivity extends AppCompatActivity {

    ArrayList<View> includedDecks;
    ArrayList<View> optionalDecks;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        optionalDecks = new ArrayList<>();
        includedDecks = new ArrayList<>();
    }

    public void onClickMonsterDeck(View v){
        if (optionalDecks.contains(v)){
            optionalDecks.remove(v);
            includedDecks.add(v);
            updateVisual(v);
        }else if (includedDecks.contains(v)){
            includedDecks.remove(v);
            updateVisual(v);
        }else{
            optionalDecks.add(v);
            updateVisual(v);
        }
    }

    private void updateVisual(View v){
        GradientDrawable gd = new GradientDrawable();
        gd.setShape(GradientDrawable.OVAL);
        if (optionalDecks.contains(v)){
            gd.setColor(Color.TRANSPARENT);
            v.setAlpha((float) 1.0);
            v.setBackground(gd);
        }else if (includedDecks.contains(v)){
            v.setAlpha((float) 1.0);
            gd.setColor(Color.BLACK);
            v.setBackground(gd);
        }else{
            clearVisual(v);
        }
    }

    private void clearVisual(View v){
        GradientDrawable gd = new GradientDrawable();
        gd.setShape(GradientDrawable.OVAL);
        v.setAlpha((float) 0.5);
        gd.setColor(Color.TRANSPARENT);
        v.setBackground(gd);
    }


    public void generate(View v){
        if (optionalDecks.size() + includedDecks.size() < 3){
            Toast.makeText(getApplicationContext(),
                    R.string.notEnoughToast,
                    Toast.LENGTH_LONG).show();
        }else if(includedDecks.size() > 4) {
            Toast.makeText(getApplicationContext(),
                    R.string.tooMuchToast,
                    Toast.LENGTH_LONG).show();
        }else{
            Intent scenario = new Intent();
            scenario.setClass(this, ScenarioActivity.class);

            ArrayList<View> chosenDecks = new ArrayList<>(includedDecks);

            if (chosenDecks.size() != 4) {
                Random r = new Random();
                Collections.shuffle(optionalDecks);



                if (r.nextBoolean() && optionalDecks.size() + includedDecks.size() > 3) {
                    chosenDecks.addAll(optionalDecks.subList(0, 4 - chosenDecks.size()));
                } else if(chosenDecks.size() != 3){
                    chosenDecks.addAll(optionalDecks.subList(0, 3 -chosenDecks.size()));
                }
            }

            scenario.putStringArrayListExtra("scenario_1",scenarioPicker(chosenDecks,1));
            scenario.putStringArrayListExtra("scenario_2",scenarioPicker(chosenDecks,2));
            scenario.putStringArrayListExtra("scenario_3",scenarioPicker(chosenDecks,3));
            scenario.putStringArrayListExtra("monster_decks",viewToString(chosenDecks));

            startActivity(scenario);
        }
    }

    private ArrayList<String> viewToString(ArrayList<View> inViews){
        ArrayList<String> outStrings = new ArrayList<>();
        for (View v : inViews){
            outStrings.add(v.getContentDescription().toString());
        }
        return outStrings;
    }


    private ArrayList<String> scenarioPicker(ArrayList<View> decks, int scenarioNumber){

        ArrayList<ArrayList<String>> options = new ArrayList<>();
        ArrayList<String> result;

        for (View v : decks) {
            ArrayList<ArrayList<String>> subOptions = new ArrayList<>();
            String deck = v.getContentDescription().toString();
            int holderId = getResources().getIdentifier(deck + "_" + Integer.toString(scenarioNumber),
                    "array",
                    this.getPackageName());
            TypedArray ta = getResources().obtainTypedArray(holderId);

            int n = ta.length();
            for (int i = 0; i < n; ++i) {
                int id = ta.getResourceId(i, 0);
                if (id > 0) {
                    subOptions.add(new ArrayList<>(Arrays.asList(getResources().getStringArray(id))));
                }
            }
            ta.recycle(); // Important!

            // zorgt ervoor dat alle encounter sets een evengrote kans hebben om gekozen te worden
            Collections.shuffle(subOptions);
            options.add(subOptions.get(0));
        }
        if (options.size() == 0){
            result = new ArrayList<>(Arrays.asList(getResources().getStringArray(R.array.general_quest)));
        }else {
            Collections.shuffle(options);
            result = options.get(0);
        }

        result.set(3,generatePoints(result.get(3)));
        return result;
    }

    private String generatePoints(String init){
        String result;
        Random r = new Random();
        int old = Integer.parseInt(init);
        result = Integer.toString(old - old / 3 + r.nextInt((old / 3 * 2) + 1));
        return result;
    }

    public void showInfo(View v){
        Intent i = new Intent(MainActivity.this, InformationPopUp.class);
        i.putExtra("source",v.getTag().toString());
        startActivity(i);
    }

    public void deselectAll(View input){
        for (View v : includedDecks){
            clearVisual(v);
        }
        for (View v : optionalDecks){
            clearVisual(v);
        }
        includedDecks.clear();
        optionalDecks.clear();
    }

    public void selectAll(View input){
        ViewGroup grid = findViewById(R.id.button_grid);
        ArrayList<View> deckButtons = getViewsByTag(grid, "image_button");
        includedDecks.clear();
        optionalDecks = deckButtons;
        for (View deck : optionalDecks){
            updateVisual(deck);
        }

    }


    private static ArrayList<View> getViewsByTag(ViewGroup root, String tag){
        ArrayList<View> views = new ArrayList<>();
        final int childCount = root.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = root.getChildAt(i);
            if (child instanceof ViewGroup) {
                views.addAll(getViewsByTag((ViewGroup) child, tag));
            }

            final Object tagObj = child.getTag();
            if (tagObj != null && tagObj.equals(tag)) {
                views.add(child);
            }

        }
        return views;
    }
}

File: app/src/main/java/com/lotrapp/sander/lotrscenariogenerator/MainActivity.kt
package com.lotrapp.sander.lotrscenariogenerator

import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import java.util.*

class MainActivity : AppCompatActivity() {

    private var includedDecks: ArrayList<View> = ArrayList()
    private var optionalDecks: ArrayList<View> = ArrayList()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    fun onClickMonsterDeck(v: View) {
        when {
            optionalDecks.contains(v) -> {
                optionalDecks.remove(v)
                includedDecks.add(v)
                updateVisual(v)
            }
            includedDecks.contains(v) -> {
                includedDecks.remove(v)
                updateVisual(v)
            }
            else -> {
                optionalDecks.add(v)
                updateVisual(v)
            }
        }
    }

    private fun updateVisual(v: View) {
        val gd = GradientDrawable()
        gd.shape = GradientDrawable.OVAL
        when {
            optionalDecks.contains(v) -> {
                gd.setColor(Color.TRANSPARENT)
                v.alpha = 1.0.toFloat()
                v.background = gd
            }
            includedDecks.contains(v) -> {
                v.alpha = 1.0.toFloat()
                gd.setColor(Color.BLACK)
                v.background = gd
            }
            else -> clearVisual(v)
        }
    }

    private fun clearVisual(v: View) {
        val gd = GradientDrawable()
        gd.shape = GradientDrawable.OVAL
        v.alpha = 0.5.toFloat()
        gd.setColor(Color.TRANSPARENT)
        v.background = gd
    }


    fun generate(@Suppress("UNUSED_PARAMETER") v: View) {
        when {
            optionalDecks.size + includedDecks.size < 3 -> Toast.makeText(applicationContext,
                    R.string.notEnoughToast,
                    Toast.LENGTH_LONG).show()
            includedDecks.size > 4 -> Toast.makeText(applicationContext,
                    R.string.tooMuchToast,
                    Toast.LENGTH_LONG).show()
            else -> {
                val scenario = Intent()
                scenario.setClass(this, QuestActivity::class.java)

                val chosenDecks = ArrayList(includedDecks)

                if (chosenDecks.size != 4) {
                    val r = Random()
                    optionalDecks.shuffle()



                    if (r.nextBoolean() && optionalDecks.size + includedDecks.size > 3) {
                        chosenDecks.addAll(optionalDecks.subList(0, 4 - chosenDecks.size))
                    } else if (chosenDecks.size != 3) {
                        chosenDecks.addAll(optionalDecks.subList(0, 3 - chosenDecks.size))
                    }
                }
                for (i in 0..2){
                    scenarioPicker(chosenDecks, i)
                }
                scenario.putStringArrayListExtra("monster_decks", viewToString(chosenDecks))

                startActivity(scenario)
            }
        }
    }

    private fun viewToString(inViews: ArrayList<View>): ArrayList<String> {
        val outStrings = ArrayList<String>()
        for (v in inViews) {
            outStrings.add(v.contentDescription.toString())
        }
        return outStrings
    }


    private fun scenarioPicker(decks: ArrayList<View>, scenarioNumber: Int) {

        val options = ArrayList<QuestCard>()
        val chosenQuestCard: QuestCard

        for (v in decks) {

            val deck = v.contentDescription.toString()

            val subOptions = when (scenarioNumber) {
                0 -> DeckStack.encounterDecks[deck]?.scenarios1
                1 -> DeckStack.encounterDecks[deck]?.scenarios2
                2 -> DeckStack.encounterDecks[deck]?.scenarios3
                else -> null
            }
            subOptions?.shuffle()
            options.add(subOptions!![0])
        }
        chosenQuestCard = if (options.size == 0) {
            DeckStack.scenarios["general_quest"]!!
        } else {
            options.shuffle()
            options[0]
        }
        if (Scenario.questCards.lastIndex < scenarioNumber){
            Scenario.questCards.add(chosenQuestCard)
            Scenario.victoryPoints.add(generatePoints(chosenQuestCard.victoryPoints))
        }else {
            Scenario.questCards[scenarioNumber] = chosenQuestCard
            Scenario.victoryPoints[scenarioNumber] = generatePoints(chosenQuestCard.victoryPoints)
        }
    }

    private fun generatePoints(init: Int): Int {
        val r = Random()
        return (init - init / 3 + r.nextInt(init / 3 * 2 + 1))
    }

    fun showInfo(v: View) {
        val i = Intent(this@MainActivity, InformationPopUp::class.java)
        i.putExtra("source", v.tag.toString())
        startActivity(i)
    }

    fun deselectAll(@Suppress("UNUSED_PARAMETER") input: View) {
        for (v in includedDecks) {
            clearVisual(v)
        }
        for (v in optionalDecks) {
            clearVisual(v)
        }
        includedDecks.clear()
        optionalDecks.clear()
    }

    fun selectAll(@Suppress("UNUSED_PARAMETER") input: View) {
        val grid = findViewById<ViewGroup>(R.id.button_grid)
        val deckButtons = getViewsByTag(grid, "image_button")
        includedDecks.clear()
        optionalDecks = deckButtons
        for (deck in optionalDecks) {
            updateVisual(deck)
        }

    }


    private fun getViewsByTag(root: ViewGroup, tag: String?): ArrayList<View> {
        val views = ArrayList<View>()
        val childCount = root.childCount
        for (i in 0 until childCount) {
            val child = root.getChildAt(i)
            if (child is ViewGroup) {
                views.addAll(getViewsByTag(child, tag))
            }

            val tagObj = child.tag
            if (tagObj != null && tagObj == tag) {
                views.add(child)
            }

        }
        return views
    }
}
